src/field_theory/laurent.lean

```diff
@@ -0,0 +1,158 @@
+-/
+
+universe u
+namespace ratfunc
+noncomputable theory
+open polynomial
+open_locale classical
+
+variables {R : Type u} [comm_ring R] [hdomain : is_domain R]
+  (r s : R) (p q : polynomial R) (f : ratfunc R)
+
+lemma taylor_non_zero_divisors (hp : p ∈ non_zero_divisors (polynomial R)) :
+  taylor r p ∈ non_zero_divisors (polynomial R) :=
+begin
+  rw mem_non_zero_divisors_iff,
+  intros x hx,
+  have : x = taylor (r - r) x,
+  { simp },
+  have ht := polynomial.taylor_injective r,
+  rwa [this, sub_eq_add_neg, ←taylor_taylor, ←taylor_mul,
+       linear_map.map_eq_zero_iff _ (taylor_injective _),
+       mul_right_mem_non_zero_divisors_eq_zero_iff hp,
+       linear_map.map_eq_zero_iff _ (taylor_injective _)] at hx,
+end
+
+/-- The Laurent expansion of rational functions about a value.
+Auxiliary definition, usage should prefer `ratfunc.laurent`. -/
+def laurent_aux : ratfunc R → ratfunc R :=
+λ f, ratfunc.lift_on f
+  (λ p q, if hq : q ∈ non_zero_divisors (polynomial R) then
+    of_fraction_ring (localization.mk (taylor r p)
+      ⟨(taylor r q), taylor_non_zero_divisors _ _ hq⟩) else 0)
+  begin
+    intros p q p' q' hq hq' h,
+    simp only [hq, hq', dif_pos, localization.mk_eq_mk', is_localization.mk'_eq_iff_eq,
+               subtype.coe_mk, h, ←taylor_mul]
+  end
+
+lemma laurent_aux_of_fraction_ring_mk (q : non_zero_divisors (polynomial R)) :
+  laurent_aux r (of_fraction_ring (localization.mk p q)) =
+    of_fraction_ring (localization.mk (taylor r p)
+      ⟨taylor r q, taylor_non_zero_divisors r q q.prop⟩) :=
+begin
+  dsimp [laurent_aux],
+  rw [lift_on_of_fraction_ring_mk, dif_pos]
+end
+
+include hdomain
+
+lemma laurent_aux_div :
+  laurent_aux r (algebra_map _ _ p / (algebra_map _ _ q)) =
+    algebra_map _ _ (taylor r p) / (algebra_map _ _ (taylor r q)) :=
+begin
+  dsimp [laurent_aux],
+  rw [lift_on_div],
+  { split_ifs with hq hq,
+    { rw [localization.mk_eq_mk', ←mk_coe_def, mk_eq_div, subtype.coe_mk] },
+    { rw [mem_non_zero_divisors_iff_ne_zero, not_not] at hq,
+      simp only [hq, div_zero, _root_.map_zero] } },
+  { simp only [div_zero, if_t_t, dite_eq_ite, is_localization.alg_equiv_apply, _root_.map_zero,
+               is_localization.ring_equiv_of_ring_equiv_apply, set_like.coe_mk, forall_const,
+               ring_equiv.to_fun_eq_coe, zero_div, of_fraction_ring_eq, fraction_ring.mk_eq_div] },
+  { intros p q p' q' hq hq' h,
+    rw ←mem_non_zero_divisors_iff_ne_zero at hq hq',
+    rw [dif_pos hq, dif_pos hq', of_fraction_ring_injective.eq_iff, localization.mk_eq_mk_iff],
+    refine localization.r_of_eq _,
+    simp only [subtype.coe_mk, ←taylor_mul, h] }
+end
+
+@[simp] lemma laurent_aux_algebra_map :
+  laurent_aux r (algebra_map _ _ p) = algebra_map _ _ (taylor r p) :=
+by rw [←mk_one, ←mk_one, mk_eq_div, laurent_aux_div, mk_eq_div, taylor_one, _root_.map_one]
+
+/-- The Laurent expansion of rational functions about a value. -/
+def laurent : ratfunc R →ₐ[R] ratfunc R :=
```

Comment:
Can this be defined as the composition of the coercion `ratfunc → laurent` and the map `ratfunc → ratfunc` that translates the coordinate by `r`?

Reply:
I can define a general `ratfunc.comp` similar to `polynomial.comp` and have a specialization of it for a translation.

This `laurent` is a bad name because Laurent is an overloaded term here. Here, I am not referring to the embedding of rational functions into Laurent series. Rather, it is the Laurent expansion, which is the rational function equivalent of the Taylor expansion for polynomials. Should this function just be called `taylor` too? Similar to how we don't have Taylor expansions of arbitrary functions, just of polynomials, I was hoping to have a Laurent expansion for rational functions to easily get proofs about roots and poles.

Happy to take comments on the usefulness of such API.

Reply:
I couldn't define it like this because `ratfunc.comp` is clunkier, since one has to carry around a proof that the denominator won't become zero. Do you still want me to build that API to show `laurent r f = f.comp (X + C r)`?

Reply:
Nah, let's leave that for future work